components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GlobalQualifiersController.java

```diff
@@ -112,31 +119,31 @@
         return null;
     }
 
-    @SuppressWarnings("unchecked")
     @Override
     public void save(Patient patient, DocumentModelBridge doc)
     {
-
         PatientData<List<VocabularyTerm>> data = patient.getData(this.getName());
 
-        BaseObject dataHolder = ((XWikiDocument) doc).getXObject(Patient.CLASS_REFERENCE);
+        BaseObject dataHolder =
+            ((XWikiDocument) doc).getXObject(Patient.CLASS_REFERENCE, true, this.xcontextProvider.get());
         if (data == null || dataHolder == null) {
             return;
         }
+        BaseClass xclass = dataHolder.getXClass(this.xcontextProvider.get());
         for (String propertyName : getProperties()) {
             List<VocabularyTerm> terms = data.get(propertyName);
             if (terms == null) {
                 continue;
             }
-            BaseProperty<ObjectPropertyReference> field =
-                (BaseProperty<ObjectPropertyReference>) dataHolder.getField(propertyName);
-            if (field != null) {
-                String fieldType = field.getClassType();
-                if (StringUtils.equals(fieldType, "com.xpn.xwiki.objects.StringProperty")) {
-                    /* there should be only one term present; just taking the head of the list */
-                    field.setValue(terms.isEmpty() ? null : termsToXWikiFormat(terms).get(0));
-                } else if (StringUtils.equals(fieldType, "com.xpn.xwiki.objects.DBStringListProperty")) {
-                    ((DBStringListProperty) field).setList(termsToXWikiFormat(terms));
+            PropertyClass xpropertyClass = (PropertyClass) xclass.get(propertyName);
+            if (xpropertyClass != null) {
+                PropertyInterface xproperty = xpropertyClass.newProperty();
+                if (xproperty instanceof BaseStringProperty) {
+                    // there should be only one term present; just taking the head of the list
+                    dataHolder.set(propertyName, terms.isEmpty() ? null : termsToXWikiFormat(terms).get(0),
+                        this.xcontextProvider.get());
+                } else if (xproperty instanceof ListProperty) {
+                    dataHolder.set(propertyName, termsToXWikiFormat(terms), this.xcontextProvider.get());
```

Comment:
`this.xcontextProvider.get()` mentioned 4 times, may be it looks better to `XWikiContext context = this.xcontextProvider.get();` and then use `context` all over the lines